File: server/checkers/game.py
from checkers.board import Board
import time

class Game:
    def __init__(self):
        self._init()
    
    def __init__(self):
        self.selected = None
        self.board = Board()
        self.turn = "WHITE"
        self.valid_moves = {}
        self.time_left = {
            "WHITE": 300, 
            "BLACK": 300
        }
        self.turn_start_time = time.time()
        
    def winner(self):
        return self.board.winner()

    def reset(self):
        self._init()

    def select(self, row, col):
        if self.selected:
            result = self._move(row, col)
            if not result:
                self.selected = None
                self.select(row, col)
        
        piece = self.board.get_piece(row, col)
        if piece != 0 and piece.color == self.turn:
            self.selected = piece
            self.valid_moves = self.board.get_valid_moves(piece)
            return True
            
        return False

    def _move(self, row, col):
        piece = self.board.get_piece(row, col)
        if self.selected and piece == 0 and (row, col) in self.valid_moves:
            self.board.move(self.selected, row, col)
            skipped = self.valid_moves[(row, col)]
            if skipped:
                self.board.remove(skipped)
            self.change_turn()
        else:
            return False

        return True

    def change_turn(self):
        self.valid_moves = {}
        
        elapsed_time = time.time() - self.turn_start_time
        time_left = self.time_left[self.turn] - int(elapsed_time)
        
        self.time_left[self.turn] = max(time_left, 0)
        
        if self.turn == "BLACK":
            self.turn = "WHITE"
        else:
            self.turn = "BLACK"
            
        self.turn_start_time = time.time() 

    def ai_move(self, board):
        self.board = board
        self.change_turn()

    def get_game_state(self):
        return {
            "board": self.board.serialized,         
            "turn": self.turn,                      
            "selected_piece": self._piece_to_json(self.selected),
            "valid_moves": list(self.valid_moves.keys()), 
            "winner": self.board.winner(),
            "time_left": self.time_left
        }

    def _piece_to_json(self, piece):
        if piece is None:
            return None
        return {
            "row": piece.row,
            "col": piece.col,
            "color": piece.color,
            "king": piece.king
        }

File: client/app/game/game.ts
import { Board } from './board';
import { Piece, Color } from './piece';

interface TimeLeft {
  WHITE: number;
  BLACK: number;
}

interface GameState {
  board: { color: Color | null; king?: boolean }[][];
  turn: Color;
  selected_piece: {
    row: number;
    col: number;
    color: Color;
    king: boolean;
  } | null;
  valid_moves: [number, number][];
  winner: Color | null;
  time_left: TimeLeft;
}

export class Game {
  private selected: Piece | null;
  private board: Board;
  private turn: Color;
  private valid_moves: Record<string, Piece[]>;
  private time_left: TimeLeft;
  private turn_start_time: number;
  private _isPaused: boolean;

  public get isPaused(): boolean {
    return this._isPaused;
  }

  constructor(turnDuration: number) {
    this.selected = null;
    this.board = new Board();
    this.turn = "WHITE";
    this.valid_moves = {};
    this.time_left = { 
      WHITE: turnDuration, 
      BLACK: turnDuration 
    };
    this.turn_start_time = Date.now() / 1000;
    this._isPaused = false;
  }

  winner(): Color | null {
    return this.board.winner();
  }

  reset(turnDuration?: number): void {
    this.selected = null;
    this.board = new Board();
    this.turn = "WHITE";
    this.valid_moves = {};
    this.time_left = { 
      WHITE: turnDuration || this.time_left.WHITE, 
      BLACK: turnDuration || this.time_left.BLACK 
    };
    this.turn_start_time = Date.now() / 1000;
    this._isPaused = false;
  }

  clear(): void {
    this.selected = null;
    this.board = new Board();
    this.turn = "WHITE";
    this.valid_moves = {};
    this._isPaused = false;
    this.turn_start_time = Date.now() / 1000;
    // Change this line to use the initial turnDuration
    this.time_left = { 
        WHITE: this.time_left.WHITE,  // Keep existing time instead of 300
        BLACK: this.time_left.BLACK 
    };
  }

  select(row: number, col: number): boolean {
    if (this.selected) {
      const moved = this._move(row, col);
      if (!moved) {
        this.selected = null;
        return this.select(row, col);
      }
    }

    const piece = this.board.get_piece(row, col);
    if (piece !== null && piece.color === this.turn) {
      this.selected = piece;
      this.valid_moves = this.board.get_valid_moves(piece);
      return true;
    }

    return false;
  }

  private _move(row: number, col: number): boolean {
    const piece = this.board.get_piece(row, col);
    const key = `${row},${col}`;
    if (this.selected && piece === null && key in this.valid_moves) {
      this.board.move(this.selected, row, col);
      const skipped = this.valid_moves[key];
      if (skipped.length > 0) this.board.remove(skipped);
      this.change_turn();
      return true;
    }
    return false;
  }

  pause(): void {
    if (!this._isPaused) {
      const now = Date.now() / 1000;
      const elapsed = now - this.turn_start_time;
      this.time_left[this.turn] = Math.max(0, this.time_left[this.turn] - Math.floor(elapsed));
      this._isPaused = true;
    }
  }

  resume(): void {
    if (this._isPaused) {
      this.turn_start_time = Date.now() / 1000;
      this._isPaused = false;
    }
  }

  private change_turn(): void {
    this.valid_moves = {};
    if (!this._isPaused) {
      const now = Date.now() / 1000;
      const elapsed = now - this.turn_start_time;
      this.time_left[this.turn] = Math.max(0, this.time_left[this.turn] - Math.floor(elapsed));
      this.turn = this.turn === 'WHITE' ? 'BLACK' : 'WHITE';
      this.turn_start_time = now;
    }
  }

  ai_move(board: Board): void {
    this.board = board;
    this.change_turn();
  }

  get_board(): Board {
    return this.board;
  }
  
  get_game_state(): GameState {
    return {
      board: this.board.serialized,
      turn: this.turn,
      selected_piece: this.selected ? {
        row: this.selected.row,
        col: this.selected.col,
        color: this.selected.color,
        king: this.selected.king
      } : null,
      valid_moves: Object.keys(this.valid_moves).map(k => k.split(',').map(Number) as [number, number]),
      winner: this.board.winner(),
      time_left: this.time_left
    };
  }
}

export default Game;
